source/common/network/socket_interface_factory.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include "envoy/config/typed_config.h"
+#include "envoy/network/socket.h"
+#include "envoy/registry/registry.h"
+
+#include "common/singleton/threadsafe_singleton.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Network {
+
+class SocketInterfaceFactory : public virtual Config::TypedFactory {
+public:
+  virtual SocketInterfacePtr createSocketInterface(const Protobuf::Message& config) PURE;
+  std::string category() const override { return "envoy.config.core.socket_interface"; }
+};
+
+class DefaultSocketInterfaceFactory : public SocketInterfaceFactory {
+public:
+  SocketInterfacePtr createSocketInterface(const Protobuf::Message& config) override;
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override;
+  std::string name() const override { return "envoy.config.core.default_socket_interface"; }
+};
+
+DECLARE_FACTORY(DefaultSocketInterfaceFactory);
+
+using SocketInterfacesMap = absl::flat_hash_map<std::string, SocketInterfacePtr>;
+using SocketInterfacesSingleton = InjectableSingleton<SocketInterfacesMap>;
+
+static inline const SocketInterface* socketInterface(std::string name) {
```

Comment:
This is to be used to find the right `SocketInterface` implementation when `Address::Instance` has a non-empty `socket_interface` field.

Reply:
The way I set this up, and maybe that's not what we want, is: typed_config is handed off to `SocketInterfaceFactory` registered via `REGISTER_FACTORY` (eg declaration above at line 27 is registered [here](https://github.com/envoyproxy/envoy/pull/11630/files/cc4d07554d7c3e884d2b49328754947d8627593d#diff-8a5c90e5a39b2ea975170edc4434345bR93)), and looked up via typed_config.name(), which then creates a `SocketInterface` implementation that we add to the hash table stored in the `SocketInterfacesSingleton`. So this new hash table is for implementations instances as opposed to factories. 

If there's a nicer way this could be done with type_url, point me to an example and I'll try to follow it (still have a lot of things to learn :-() .

Reply:
I will defer to @lizan and @kyessenov who have more experience with the factory, but I'm wondering if we could change it slightly so the registered factory just spits out sockets and then we would just look up the factory b ytype_url?